File: packages/shared-types/src/analysis.ts
/** Mirrors `app/schemas/analysis.py` and `app/enums.py`. */

export type AnalysisType = "image" | "text";
export type AnalysisStatus = "queued" | "processing" | "completed" | "failed";
export type StepStatus = "running" | "completed" | "failed" | "skipped";

/** One independently identifiable pipeline step (mirrors `AnalysisStepResponse`). */
export interface AnalysisStepResponse {
  name: string;
  status: StepStatus;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number | null;
  error_code: string | null;
  error_message: string | null;
  details: Record<string, unknown> | null;
}

/** Client-safe view of the stored original; storage keys are never exposed. */
/** Short-lived signed URL for the stored original (owner only). */
export interface AnalysisFileLink {
  url: string;
  expires_in_seconds: number;
  mime_type: string | null;
  width: number | null;
  height: number | null;
}

export interface AnalysisFileResponse {
  original_filename: string | null;
  mime_type: string | null;
  size_bytes: number | null;
  sha256: string;
  width: number | null;
  height: number | null;
}

export interface AnalysisResponse {
  id: string;
  type: AnalysisType;
  status: AnalysisStatus;
  title: string | null;
  error_code: string | null;
  error_message: string | null;
  created_at: string;
  completed_at: string | null;
  /** Raw-content expiry (null = never), owner keep flag, and when content was purged. */
  retention_at: string | null;
  kept_at: string | null;
  content_purged_at: string | null;
  file: AnalysisFileResponse | null;
  steps: AnalysisStepResponse[];
}

export interface AnalysisCreatedResponse {
  id: string;
  status: AnalysisStatus;
  type: AnalysisType;
}

export interface AnalysisListResponse {
  items: AnalysisResponse[];
  page: number;
  page_size: number;
  total: number;
}

export interface AnalysisCounts {
  total: number;
  image: number;
  text: number;
}

/** Upload constraints enforced by the API (client checks are a courtesy only). */
export const IMAGE_UPLOAD = {
  acceptedMimeTypes: ["image/jpeg", "image/png", "image/webp", "image/tiff"],
  maxBytes: 25 * 1024 * 1024,
} as const;

File: services/api/app/schemas/analysis.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.enums import AnalysisStatus, AnalysisType

MAX_TITLE_LENGTH = 300


class AnalysisFileResponse(BaseModel):
    """Client-safe view of the stored original. The storage key is never exposed."""

    model_config = ConfigDict(from_attributes=True)

    original_filename: str | None
    mime_type: str | None
    size_bytes: int | None
    sha256: str
    width: int | None
    height: int | None


class AnalysisResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    type: AnalysisType
    status: AnalysisStatus
    title: str | None
    error_code: str | None
    error_message: str | None
    created_at: datetime
    completed_at: datetime | None
    file: AnalysisFileResponse | None = None


class AnalysisCreatedResponse(BaseModel):
    """Matches the API contract for creation endpoints."""

    id: uuid.UUID
    status: AnalysisStatus
    type: AnalysisType


class AnalysisListResponse(BaseModel):
    items: list[AnalysisResponse]
    page: int = Field(ge=1)
    page_size: int = Field(ge=1)
    total: int = Field(ge=0)


class AnalysisCounts(BaseModel):
    total: int
    image: int
    text: int
